Create Fabric compatibility issue - Patch F

## Crash
(workaround below)
When playing with Cobblegen (any version) and Create Fabric installed, the game crashes with an error like this one:
```abap
[15/03/2024 20:41:33 PM] ---- Minecraft Crash Report ----
[15/03/2024 20:41:33 PM] // Hey, that tickles! Hehehe!
[15/03/2024 20:41:33 PM] 
[15/03/2024 20:41:33 PM] Time: 2024-03-15 20:41:33
[15/03/2024 20:41:33 PM] Description: Initializing game
[15/03/2024 20:41:33 PM] 
[15/03/2024 20:41:33 PM] java.lang.RuntimeException: Could not execute entrypoint stage 'main' due to errors, provided by 'create'!
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.FabricLoaderImpl.lambda$invokeEntrypoints$2(FabricLoaderImpl.java:388)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.util.ExceptionUtil.gatherExceptions(ExceptionUtil.java:33)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.FabricLoaderImpl.invokeEntrypoints(FabricLoaderImpl.java:386)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.game.minecraft.Hooks.startClient(Hooks.java:52)
[15/03/2024 20:41:33 PM] 	at net.minecraft.class_310.<init>(class_310.java:458)
[15/03/2024 20:41:33 PM] 	at net.minecraft.client.main.Main.main(Main.java:211)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:470)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23)
[15/03/2024 20:41:33 PM] Caused by: java.lang.BootstrapMethodError: java.lang.RuntimeException: Mixin transformation of com.simibubi.create.content.fluids.FluidReactions failed
[15/03/2024 20:41:33 PM] 	at com.simibubi.create.foundation.events.CommonEvents.register(CommonEvents.java:229)
[15/03/2024 20:41:33 PM] 	at com.simibubi.create.Create.onInitialize(Create.java:154)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.FabricLoaderImpl.invokeEntrypoints(FabricLoaderImpl.java:384)
[15/03/2024 20:41:33 PM] 	... 6 more
[15/03/2024 20:41:33 PM] Caused by: java.lang.RuntimeException: Mixin transformation of com.simibubi.create.content.fluids.FluidReactions failed
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:427)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:323)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:218)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:119)
[15/03/2024 20:41:33 PM] 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
[15/03/2024 20:41:33 PM] 	... 9 more
[15/03/2024 20:41:33 PM] Caused by: org.spongepowered.asm.mixin.transformer.throwables.MixinTransformerError: An unexpected critical error was encountered
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:392)
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234)
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:422)
[15/03/2024 20:41:33 PM] 	... 13 more
[15/03/2024 20:41:33 PM] Caused by: org.spongepowered.asm.mixin.transformer.throwables.MixinPreProcessorException: Attach error for cobblegen.mixins.json:fluid.CreateFluidReactionsMixin from mod cobblegen during activity: [Transform -> Method generator$handlePipeFlowCollision(Lnet/minecraft/class_1937;Lnet/minecraft/class_2338;Lio/github/fabricators_of_create/porting_lib/util/FluidStack;Lio/github/fabricators_of_create/porting_lib/util/FluidStack;Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfo;)V -> INVOKEVIRTUAL -> io/github/fabricators_of_create/porting_lib/util/FluidStack::getFluid:()Lnet/minecraft/class_3611;]
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.attach(MixinPreProcessorStandard.java:313)
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.createContextFor(MixinPreProcessorStandard.java:277)
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinInfo.createContextFor(MixinInfo.java:1289)
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:294)
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.TargetClassContext.apply(TargetClassContext.java:421)
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:403)
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:363)
[15/03/2024 20:41:33 PM] 	... 16 more
[15/03/2024 20:41:33 PM] Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: io.github.fabricators_of_create.porting_lib.util.FluidStack
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.transformMemberReference(MixinPreProcessorStandard.java:786)
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.transformMethod(MixinPreProcessorStandard.java:772)
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.transform(MixinPreProcessorStandard.java:738)
[15/03/2024 20:41:33 PM] 	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.attach(MixinPreProcessorStandard.java:307)
[15/03/2024 20:41:33 PM] 	... 22 more
[15/03/2024 20:41:33 PM] Caused by: java.lang.ClassNotFoundException: io.github.fabricators_of_create.porting_lib.util.FluidStack
[15/03/2024 20:41:33 PM] 	... 26 more
[15/03/2024 20:41:33 PM] 
[15/03/2024 20:41:33 PM] 
[15/03/2024 20:41:33 PM] A detailed walkthrough of the error, its code path and all known details is as follows:
[15/03/2024 20:41:33 PM] ---------------------------------------------------------------------------------------
[15/03/2024 20:41:33 PM] 
[15/03/2024 20:41:33 PM] -- Head --
[15/03/2024 20:41:33 PM] Thread: Render thread
[15/03/2024 20:41:33 PM] Stacktrace:
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.FabricLoaderImpl.lambda$invokeEntrypoints$2(FabricLoaderImpl.java:388)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.util.ExceptionUtil.gatherExceptions(ExceptionUtil.java:33)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.FabricLoaderImpl.invokeEntrypoints(FabricLoaderImpl.java:386)
[15/03/2024 20:41:33 PM] 	at net.fabricmc.loader.impl.game.minecraft.Hooks.startClient(Hooks.java:52)
[15/03/2024 20:41:33 PM] 	at net.minecraft.class_310.<init>(class_310.java:458)
```

## Known Workaround:  
(while a fix is being created) 
- Navigate to your mods folder
- Rename the cobblestone jar from `.jar` to `.zip`
- Open its contents and find the file called "`cobblegen.mixins.json`"
- Open the file in a text editor
- Remove all of line 9 (`    "fluid.CreateFluidReactionsMixin",`)
- When you're done, it should look like this:
![image](https://github.com/null2264/CobbleGen/assets/33235902/dff290ef-9f92-4137-8c94-7e4852cfcfe7)
- Save the file
- Rename the file back from `.zip` to `.jar`
- Note: this workaround likely disables all create contraption interactions! Use with caution.

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -9,12 +9,12 @@ jobs:
       - uses: actions/checkout@v4
       - uses: gradle/wrapper-validation-action@v2
       - name: Setup JDK
-        uses: actions/setup-java@v3
+        uses: actions/setup-java@v4
         with:
           distribution: temurin
           java-version: 17
       - run: chmod +x ./gradlew
-      - uses: actions/cache@v3
+      - uses: actions/cache@v4
         with:
           path: ~/.gradle/caches
           key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*', 'gradle.properties') }}
@@ -25,7 +25,7 @@ jobs:
       - name: Archive build results
         run: tar -I zstd -cf build.tar.zst versions/*/build/libs
       - name: Upload build folder
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           name: build-artifacts
           path: build.tar.zst

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -7,7 +7,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v4
-      - uses: gradle/wrapper-validation-action@v1
+      - uses: gradle/wrapper-validation-action@v2
       - name: Setup JDK
         uses: actions/setup-java@v3
         with:

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -9,7 +9,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v4
-      - uses: gradle/wrapper-validation-action@v1
+      - uses: gradle/wrapper-validation-action@v2
       - name: Setup JDK
         uses: actions/setup-java@v3
         with:

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -11,12 +11,12 @@ jobs:
       - uses: actions/checkout@v4
       - uses: gradle/wrapper-validation-action@v2
       - name: Setup JDK
-        uses: actions/setup-java@v3
+        uses: actions/setup-java@v4
         with:
           distribution: temurin
           java-version: 17
       - run: chmod +x ./gradlew
-      - uses: actions/cache@v3
+      - uses: actions/cache@v4
         with:
           path: ~/.gradle/caches
           key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*', 'gradle.properties') }}
@@ -27,7 +27,7 @@ jobs:
       - name: Archive build results
         run: tar -I zstd -cf build.tar.zst versions/*/build/libs
       - name: Upload build and gradle folders
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           name: build-artifacts
           path: build.tar.zst
@@ -46,7 +46,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Download build artifact
-        uses: actions/download-artifact@v3
+        uses: actions/download-artifact@v4
         with:
           name: build-artifacts
       - name: Unpack build artifact

--- a/build.gradle
+++ b/build.gradle
@@ -245,10 +245,10 @@ dependencies {
 
 shadowJar {
     zip64 true
-    relocate "blue.endless.jankson", "io.github.null2264.shadowed.blue.endless.jankson"
+    relocate "blue.endless.jankson", "io.github.null2264.shadowed.jankson"
     if (mcVersion < 12000) {
-        relocate "org.slf4j", "io.github.null2264.shadowed.org.slf4j"
-        relocate "org.apache.logging.slf4j", "io.github.null2264.shadowed.org.apache.logging.slf4j"
+        relocate "org.slf4j", "io.github.null2264.shadowed.slf4j"
+        relocate "org.apache.logging", "io.github.null2264.shadowed.log4j"
     }
     relocate "manifold", "io.github.null2264.shadowed.manifold"
     exclude (isFabric ? "META-INF/mods.toml" : "fabric.mod.json")

--- a/build.gradle
+++ b/build.gradle
@@ -235,20 +235,20 @@ dependencies {
     } else {
         // slf4j is not included by MC in 1.16.5
         shade("org.slf4j:slf4j-api:1.7.36")
-        shade("org.apache.logging.log4j:log4j-slf4j-impl:2.17.0")
+        shade("org.apache.logging.log4j:log4j-slf4j-impl:2.8.1")
         if (!isFabric) {
             forgeRuntimeLibrary("org.slf4j:slf4j-api:1.7.36")
-            forgeRuntimeLibrary("org.apache.logging.log4j:log4j-slf4j-impl:2.17.0")
+            forgeRuntimeLibrary("org.apache.logging.log4j:log4j-slf4j-impl:2.8.1")
         }
     }
 }
 
 shadowJar {
     zip64 true
-    relocate "blue.endless.jankson", "io.github.null2264.shadowed.blue.endless.jankson"
+    relocate "blue.endless.jankson", "io.github.null2264.shadowed.jankson"
     if (mcVersion < 12000) {
-        relocate "org.slf4j", "io.github.null2264.shadowed.org.slf4j"
-        relocate "org.apache.logging.slf4j", "io.github.null2264.shadowed.org.apache.logging.slf4j"
+        relocate "org.slf4j", "io.github.null2264.shadowed.slf4j"
+        relocate "org.apache.logging", "io.github.null2264.shadowed.log4j"
     }
     relocate "manifold", "io.github.null2264.shadowed.manifold"
     exclude (isFabric ? "META-INF/mods.toml" : "fabric.mod.json")

--- a/build.gradle
+++ b/build.gradle
@@ -235,20 +235,20 @@ dependencies {
     } else {
         // slf4j is not included by MC in 1.16.5
         shade("org.slf4j:slf4j-api:1.7.36")
-        shade("org.apache.logging.log4j:log4j-slf4j-impl:2.8.1")
+        shade("org.apache.logging.log4j:log4j-slf4j-impl:2.17.0")
         if (!isFabric) {
             forgeRuntimeLibrary("org.slf4j:slf4j-api:1.7.36")
-            forgeRuntimeLibrary("org.apache.logging.log4j:log4j-slf4j-impl:2.8.1")
+            forgeRuntimeLibrary("org.apache.logging.log4j:log4j-slf4j-impl:2.17.0")
         }
     }
 }
 
 shadowJar {
     zip64 true
-    relocate "blue.endless.jankson", "io.github.null2264.shadowed.jankson"
+    relocate "blue.endless.jankson", "io.github.null2264.shadowed.blue.endless.jankson"
     if (mcVersion < 12000) {
-        relocate "org.slf4j", "io.github.null2264.shadowed.slf4j"
-        relocate "org.apache.logging", "io.github.null2264.shadowed.log4j"
+        relocate "org.slf4j", "io.github.null2264.shadowed.org.slf4j"
+        relocate "org.apache.logging.slf4j", "io.github.null2264.shadowed.org.apache.logging.slf4j"
     }
     relocate "manifold", "io.github.null2264.shadowed.manifold"
     exclude (isFabric ? "META-INF/mods.toml" : "fabric.mod.json")

--- a/build.gradle
+++ b/build.gradle
@@ -50,6 +50,7 @@ repositories {
     maven { url = "https://maven.tterrag.com/" }
     maven { url = "https://maven.theillusivec4.top/" }
     maven { url = "https://maven.neoforged.net/releases" }
+    maven { url = "https://raw.githubusercontent.com/Fuzss/modresources/main/maven/" }
 }
 
 archivesBaseName = project.archives_base_name
@@ -149,6 +150,9 @@ dependencies {
         // These act like a dummy, technically only here to provide their modules/packages
         if (isFabric) {
             modCompileOnly("io.github.fabricators_of_create:Porting-Lib:${project.port_lib_version_1_18_2}")
+            modCompileOnly("com.simibubi.create:create-fabric-${project.minecraft_version_1_18_2}:${project.create_version_1_18_2}")
+        } else {
+            modCompileOnly("com.simibubi.create:create-1.18.2:0.5.1.e-318:slim") { transitive = false }
         }
 
         // <- EMI

--- a/build.gradle
+++ b/build.gradle
@@ -235,10 +235,10 @@ dependencies {
     } else {
         // slf4j is not included by MC in 1.16.5
         shade("org.slf4j:slf4j-api:1.7.36")
-        shade("org.apache.logging.log4j:log4j-slf4j-impl:2.8.1")
+        shade("org.apache.logging.log4j:log4j-slf4j-impl:2.17.0")
         if (!isFabric) {
             forgeRuntimeLibrary("org.slf4j:slf4j-api:1.7.36")
-            forgeRuntimeLibrary("org.apache.logging.log4j:log4j-slf4j-impl:2.8.1")
+            forgeRuntimeLibrary("org.apache.logging.log4j:log4j-slf4j-impl:2.17.0")
         }
     }
 }

--- a/build.gradle
+++ b/build.gradle
@@ -235,10 +235,10 @@ dependencies {
     } else {
         // slf4j is not included by MC in 1.16.5
         shade("org.slf4j:slf4j-api:1.7.36")
-        shade("org.apache.logging.log4j:log4j-slf4j-impl:2.17.0")
+        shade("org.apache.logging.log4j:log4j-slf4j-impl:2.8.1")
         if (!isFabric) {
             forgeRuntimeLibrary("org.slf4j:slf4j-api:1.7.36")
-            forgeRuntimeLibrary("org.apache.logging.log4j:log4j-slf4j-impl:2.17.0")
+            forgeRuntimeLibrary("org.apache.logging.log4j:log4j-slf4j-impl:2.8.1")
         }
     }
 }

--- a/build.gradle
+++ b/build.gradle
@@ -246,7 +246,7 @@ dependencies {
 shadowJar {
     zip64 true
     relocate "blue.endless.jankson", "io.github.null2264.shadowed.jankson"
-    if (mcVersion < 12000) {
+    if (mcVersion <= 11605) {
         relocate "org.slf4j", "io.github.null2264.shadowed.slf4j"
         relocate "org.apache.logging", "io.github.null2264.shadowed.log4j"
     }

--- a/gradle.properties
+++ b/gradle.properties
@@ -21,7 +21,7 @@ recipe_viewer=none
 # Dependencies
 jankson_version=1.2.3
 emi_version=1.0.26
-manifold_version=2024.1.5
+manifold_version=2024.1.9
 
 # Version Specific Dependencies
 create_version_1_18_2=0.5.1-f-build.1333+mc1.18.2

--- a/gradle.properties
+++ b/gradle.properties
@@ -21,8 +21,8 @@ recipe_viewer=none
 # Dependencies
 jankson_version=1.2.3
 emi_version=1.0.26
-manifold_version=2024.1.3
+manifold_version=2024.1.5
 
 # Version Specific Dependencies
-create_version_1_18_2=0.5.0.i-1002+1.18.2
+create_version_1_18_2=0.5.1-f-build.1333+mc1.18.2
 port_lib_version_1_18_2=1.2.869-beta+1.18.2

--- a/src/main/java/io/github/null2264/cobblegen/FluidInteraction.java
+++ b/src/main/java/io/github/null2264/cobblegen/FluidInteraction.java
@@ -159,14 +159,14 @@ public class FluidInteraction
      * Create Pipe support
      */
     @ApiStatus.Internal
-    public boolean interactFromPipe(Level level, BlockPos pos, Fluid fluid1, Fluid fluid2) {
+    public Optional<BlockState> interactFromPipeState(Level level, BlockPos pos, Fluid fluid1, Fluid fluid2) {
         Fluid source;
         Fluid neighbour;
         List<Generator> generators = generatorMap.get(fluid1);
         if (generators == null) {
             generators = generatorMap.get(fluid2);
             if (generators == null)
-                return false;
+                return Optional.empty();
             source = fluid2;
             neighbour = fluid1;
         } else {
@@ -185,9 +185,17 @@ public class FluidInteraction
 
             final Optional<BlockState> result = generator.tryGenerate(level, pos, source.defaultFluidState(), neighbour.defaultFluidState());
             if (result.isPresent()) {
-                level.setBlockAndUpdate(pos, result.get());
-                return true;
+                return result;
             }
+        }
+        return Optional.empty();
+    }
+
+    public boolean interactFromPipe(Level level, BlockPos pos, Fluid fluid1, Fluid fluid2) {
+        final Optional<BlockState> result = interactFromPipeState(level, pos, fluid1, fluid2);
+        if (result.isPresent()) {
+            level.setBlockAndUpdate(pos, result.get());
+            return true;
         }
         return false;
     }

--- a/src/main/java/io/github/null2264/cobblegen/FluidInteraction.java
+++ b/src/main/java/io/github/null2264/cobblegen/FluidInteraction.java
@@ -1,5 +1,6 @@
 package io.github.null2264.cobblegen;
 
+import io.github.null2264.cobblegen.CobbleGen;
 import io.github.null2264.cobblegen.data.CGRegistryImpl;
 import io.github.null2264.cobblegen.data.model.CGRegistry;
 import io.github.null2264.cobblegen.data.model.Generator;
@@ -160,6 +161,8 @@ public class FluidInteraction
      */
     @ApiStatus.Internal
     public Optional<BlockState> interactFromPipeState(Level level, BlockPos pos, Fluid fluid1, Fluid fluid2) {
+        if (CobbleGen.META_CONFIG.create.disablePipe) return Optional.empty();
+
         Fluid source;
         Fluid neighbour;
         List<Generator> generators = generatorMap.get(fluid1);

--- a/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
@@ -21,9 +21,13 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
         return null;
     }
 
+    /**
+     * Checks if loaded Create mod is patch F or newer
+     * @return 1 for Patch F, 0 for not Patch E, -1 for always Patch E
+     */
     private int isPatchFOrNewer() {
         //#if FABRIC<=0
-        //$$ return -1;
+        //$$ return -1;  // Always use Patch E mixin for Forge-alike
         //#else
         //#if MC>1.16.5
         try {

--- a/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
@@ -47,7 +47,7 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
 
     @Override
     public boolean shouldApplyMixin(String targetClassName, String mixinClassName) {
-        if (mixinClassName.startsWith("CreateFluidReactionsMixin")) {
+        if (mixinClassName.contains("CreateFluidReactionsMixin")) {
             if (!LoaderCompat.isModLoaded("create")) return false;
 
             if (mixinClassName.endsWith("PatchF")) return isPatchFOrNewer() >= 1 && !CobbleGen.META_CONFIG.create.disablePipe;

--- a/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
@@ -1,6 +1,5 @@
 package io.github.null2264.cobblegen.mixin.core;
 
-import io.github.null2264.cobblegen.CobbleGen;
 import io.github.null2264.cobblegen.compat.LoaderCompat;
 import org.objectweb.asm.tree.ClassNode;
 import org.spongepowered.asm.mixin.extensibility.IMixinConfigPlugin;
@@ -50,10 +49,8 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
         if (mixinClassName.contains("CreateFluidReactionsMixin")) {
             if (!LoaderCompat.isModLoaded("create")) return false;
 
-            if (mixinClassName.endsWith("PatchF")) return isPatchFOrNewer() >= 1 && !CobbleGen.META_CONFIG.create.disablePipe;
-            if (mixinClassName.endsWith("PatchE")) return isPatchFOrNewer() <= 0 && !CobbleGen.META_CONFIG.create.disablePipe;
-
-            return !CobbleGen.META_CONFIG.create.disablePipe;
+            if (mixinClassName.endsWith("PatchF")) return isPatchFOrNewer() >= 1;
+            if (mixinClassName.endsWith("PatchE")) return isPatchFOrNewer() <= 0;
         }
         return true;
     }

--- a/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
@@ -1,5 +1,6 @@
 package io.github.null2264.cobblegen.mixin.core;
 
+import io.github.null2264.cobblegen.CobbleGen;
 import io.github.null2264.cobblegen.compat.LoaderCompat;
 import org.objectweb.asm.tree.ClassNode;
 import org.spongepowered.asm.mixin.extensibility.IMixinConfigPlugin;
@@ -20,10 +21,31 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
         return null;
     }
 
+    private boolean isPatchFOrNewer() {
+        //#if FABRIC<=0
+        //$$ return false;
+        //#else
+        try {
+            String version =
+                net.fabricmc.loader.api.FabricLoader.getInstance().getModContainer("create")
+                    .orElseThrow().getMetadata().getVersion().getFriendlyString();
+            return version.startsWith("0.5.1-f") || version.startsWith("0.5.1.f");
+        } catch (java.util.NoSuchElementException exc) {
+            return false;
+        }
+        //#endif
+    }
+
     @Override
     public boolean shouldApplyMixin(String targetClassName, String mixinClassName) {
-        if (mixinClassName.endsWith("CreateFluidReactionsMixin"))
-            return LoaderCompat.isModLoaded("create");
+        if (mixinClassName.startsWith("CreateFluidReactionsMixin")) {
+            if (!LoaderCompat.isModLoaded("create")) return false;
+
+            if (mixinClassName.endsWith("PatchF")) return isPatchFOrNewer();
+            if (mixinClassName.endsWith("PatchE")) return !isPatchFOrNewer();
+
+            return !CobbleGen.META_CONFIG.create.disablePipe;
+        }
         return true;
     }
 

--- a/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
@@ -50,8 +50,8 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
         if (mixinClassName.startsWith("CreateFluidReactionsMixin")) {
             if (!LoaderCompat.isModLoaded("create")) return false;
 
-            if (mixinClassName.endsWith("PatchF")) return isPatchFOrNewer() > 0;
-            if (mixinClassName.endsWith("PatchE")) return isPatchFOrNewer() < 1;
+            if (mixinClassName.endsWith("PatchF")) return isPatchFOrNewer() > 0 && !CobbleGen.META_CONFIG.create.disablePipe;
+            if (mixinClassName.endsWith("PatchE")) return isPatchFOrNewer() < 1 && !CobbleGen.META_CONFIG.create.disablePipe;
 
             return !CobbleGen.META_CONFIG.create.disablePipe;
         }

--- a/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
@@ -34,9 +34,11 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
         } catch (java.util.NoSuchElementException exc) {
             return 0;
         }
-        //#endif
-        //#endif
         return 0;
+        //#else
+        //$$ return 0;  // We don't support create integration for MC1.16.5
+        //#endif
+        //#endif
     }
 
     @Override

--- a/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
@@ -21,17 +21,17 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
         return null;
     }
 
-    private boolean isPatchFOrNewer() {
+    private int isPatchFOrNewer() {
         //#if FABRIC<=0
-        //$$ return false;
+        //$$ return -1;
         //#else
         try {
             String version =
                 net.fabricmc.loader.api.FabricLoader.getInstance().getModContainer("create")
                     .orElseThrow().getMetadata().getVersion().getFriendlyString();
-            return version.startsWith("0.5.1-f") || version.startsWith("0.5.1.f");
+            if (version.startsWith("0.5.1-f") || version.startsWith("0.5.1.f")) return 1;
         } catch (java.util.NoSuchElementException exc) {
-            return false;
+            return 0;
         }
         //#endif
     }
@@ -41,8 +41,8 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
         if (mixinClassName.startsWith("CreateFluidReactionsMixin")) {
             if (!LoaderCompat.isModLoaded("create")) return false;
 
-            if (mixinClassName.endsWith("PatchF")) return isPatchFOrNewer();
-            if (mixinClassName.endsWith("PatchE")) return !isPatchFOrNewer();
+            if (mixinClassName.endsWith("PatchF")) return isPatchFOrNewer() > 0;
+            if (mixinClassName.endsWith("PatchE")) return isPatchFOrNewer() < 1;
 
             return !CobbleGen.META_CONFIG.create.disablePipe;
         }

--- a/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
@@ -50,8 +50,8 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
         if (mixinClassName.startsWith("CreateFluidReactionsMixin")) {
             if (!LoaderCompat.isModLoaded("create")) return false;
 
-            if (mixinClassName.endsWith("PatchF")) return isPatchFOrNewer() > 0 && !CobbleGen.META_CONFIG.create.disablePipe;
-            if (mixinClassName.endsWith("PatchE")) return isPatchFOrNewer() < 1 && !CobbleGen.META_CONFIG.create.disablePipe;
+            if (mixinClassName.endsWith("PatchF")) return isPatchFOrNewer() >= 1 && !CobbleGen.META_CONFIG.create.disablePipe;
+            if (mixinClassName.endsWith("PatchE")) return isPatchFOrNewer() <= 0 && !CobbleGen.META_CONFIG.create.disablePipe;
 
             return !CobbleGen.META_CONFIG.create.disablePipe;
         }

--- a/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
@@ -25,6 +25,7 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
         //#if FABRIC<=0
         //$$ return -1;
         //#else
+        //#if MC>1.16.5
         try {
             String version =
                 net.fabricmc.loader.api.FabricLoader.getInstance().getModContainer("create")
@@ -33,6 +34,9 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
         } catch (java.util.NoSuchElementException exc) {
             return 0;
         }
+        //#else
+        return 0;  // We don't support create integration for MC1.16.5
+        //#endif
         //#endif
     }
 

--- a/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/core/CobbleGenMixinPlugin.java
@@ -34,10 +34,9 @@ public class CobbleGenMixinPlugin implements IMixinConfigPlugin
         } catch (java.util.NoSuchElementException exc) {
             return 0;
         }
-        //#else
-        return 0;  // We don't support create integration for MC1.16.5
         //#endif
         //#endif
+        return 0;
     }
 
     @Override

--- a/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixin.java
@@ -1,5 +1,5 @@
 //#if MC>1.16.5
-package io.github.null2264.cobblegen.mixin.fluid;
+package io.github.null2264.cobblegen.mixin.create;
 
 //#if FABRIC>=1
 import io.github.fabricators_of_create.porting_lib.util.FluidStack;
@@ -27,7 +27,7 @@ import static io.github.null2264.cobblegen.CobbleGen.FLUID_INTERACTION;
 
 @SuppressWarnings("UnresolvedMixinReference") // False positive
 @Pseudo
-@Mixin(targets = {"com.simibubi.create.content.contraptions.fluids.FluidReactions", "com.simibubi.create.content.fluids.FluidReactions"})
+@Mixin(com.simibubi.create.content.fluids.FluidReactions.class)
 public abstract class CreateFluidReactionsMixin
 {
     @Unique

--- a/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixin.java
@@ -7,13 +7,15 @@ import net.minecraft.world.level.Level;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.FluidState;
 import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Pseudo;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
 import static io.github.null2264.cobblegen.CobbleGen.FLUID_INTERACTION;
 
-@Mixin(com.simibubi.create.content.fluids.FluidReactions.class)
+@Pseudo
+@Mixin(targets = {"com.simibubi.create.content.contraptions.fluids.FluidReactions", "com.simibubi.create.content.fluids.FluidReactions"})
 public abstract class CreateFluidReactionsMixin
 {
     @Inject(

--- a/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixin.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixin.java
@@ -1,56 +1,21 @@
 //#if MC>1.16.5
 package io.github.null2264.cobblegen.mixin.create;
 
-//#if FABRIC>=1
-import io.github.fabricators_of_create.porting_lib.util.FluidStack;
-//#else
-    //#if FORGE>=2 && MC>=1.20.2
-    //$$ import net.neoforged.neoforge.fluids.FluidStack;
-    //#else
-    //$$ import net.minecraftforge.fluids.FluidStack;
-    //#endif
-//#endif
-import io.github.null2264.cobblegen.CobbleGen;
 import io.github.null2264.cobblegen.data.model.Generator;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.FluidState;
 import org.spongepowered.asm.mixin.Mixin;
-import org.spongepowered.asm.mixin.Pseudo;
-import org.spongepowered.asm.mixin.Unique;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
 import static io.github.null2264.cobblegen.CobbleGen.FLUID_INTERACTION;
 
-@SuppressWarnings("UnresolvedMixinReference") // False positive
-@Pseudo
 @Mixin(com.simibubi.create.content.fluids.FluidReactions.class)
 public abstract class CreateFluidReactionsMixin
 {
-    @Unique
-    private static boolean handleReaction(Level level, BlockPos pos, Fluid fluid1, Fluid fluid2) {
-        if (CobbleGen.META_CONFIG.create.disablePipe) return false;
-        return FLUID_INTERACTION.interactFromPipe(level, pos, fluid1, fluid2);
-    }
-
-    @Inject(
-            method = "handlePipeFlowCollision",
-            //#if FABRIC<=0
-            //$$ remap = false,
-            //#endif
-            at = @At(value = "HEAD"), cancellable = true
-    )
-    private static void generator$handlePipeFlowCollision(
-            Level level, BlockPos pos, FluidStack fluid1, FluidStack fluid2, CallbackInfo ci
-    ) {
-        final boolean success = handleReaction(level, pos, fluid1.getFluid(), fluid2.getFluid());
-        if (success)
-            ci.cancel();
-    }
-
     @Inject(
             method = "handlePipeSpillCollision",
             //#if FABRIC<=0
@@ -62,7 +27,8 @@ public abstract class CreateFluidReactionsMixin
     private static void generator$handlePipeSpillCollision(
             Level level, BlockPos pos, Fluid pipeFluid, FluidState worldFluid, CallbackInfo ci
     ) {
-        final boolean success = handleReaction(level, pos, Generator.getStillFluid(pipeFluid), worldFluid.getType());
+        final boolean success =
+            FLUID_INTERACTION.interactFromPipe(level, pos, Generator.getStillFluid(pipeFluid), worldFluid.getType());
         if (success)
             ci.cancel();
     }

--- a/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchE.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchE.java
@@ -1,7 +1,6 @@
 package io.github.null2264.cobblegen.mixin.create;
 
 //#if FABRIC>=1
-import com.simibubi.create.content.fluids.FluidReactions;
 import io.github.fabricators_of_create.porting_lib.util.FluidStack;
 //#else
     //#if FORGE>=2 && MC>=1.20.2
@@ -14,6 +13,7 @@ import io.github.fabricators_of_create.porting_lib.util.FluidStack;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.level.Level;
 import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Pseudo;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
@@ -23,7 +23,8 @@ import static io.github.null2264.cobblegen.CobbleGen.FLUID_INTERACTION;
 /**
  * Mixin for Create Fabric (pre) Patch F
  */
-@Mixin(FluidReactions.class)
+@Pseudo
+@Mixin(targets = {"com.simibubi.create.content.contraptions.fluids.FluidReactions", "com.simibubi.create.content.fluids.FluidReactions"})
 public abstract class CreateFluidReactionsMixinPatchE
 {
     @Inject(

--- a/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchE.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchE.java
@@ -1,0 +1,43 @@
+package io.github.null2264.cobblegen.mixin.create;
+
+//#if FABRIC>=1
+import com.simibubi.create.content.fluids.FluidReactions;
+import io.github.fabricators_of_create.porting_lib.util.FluidStack;
+//#else
+    //#if FORGE>=2 && MC>=1.20.2
+    //$$ import net.neoforged.neoforge.fluids.FluidStack;
+    //#else
+    //$$ import net.minecraftforge.fluids.FluidStack;
+    //#endif
+//#endif
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.level.Level;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import static io.github.null2264.cobblegen.CobbleGen.FLUID_INTERACTION;
+
+/**
+ * Mixin for Create Fabric (pre) Patch F
+ */
+@Mixin(FluidReactions.class)
+public abstract class CreateFluidReactionsMixinPatchE
+{
+    @Inject(
+        method = "handlePipeFlowCollision",
+        //#if FABRIC<=0
+        //$$ remap = false,
+        //#endif
+        at = @At(value = "HEAD"), cancellable = true
+    )
+    private static void generator$handlePipeFlowCollision(
+        Level level, BlockPos pos, FluidStack fluid1, FluidStack fluid2, CallbackInfo ci
+    ) {
+        final boolean success = FLUID_INTERACTION.interactFromPipe(level, pos, fluid1.getFluid(), fluid2.getFluid());
+        if (success)
+            ci.cancel();
+    }
+}

--- a/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchE.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchE.java
@@ -1,3 +1,4 @@
+//#if MC>1.16.5
 package io.github.null2264.cobblegen.mixin.create;
 
 //#if FABRIC>=1
@@ -42,3 +43,4 @@ public abstract class CreateFluidReactionsMixinPatchE
             ci.cancel();
     }
 }
+//#endif

--- a/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchF.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchF.java
@@ -1,0 +1,37 @@
+//#if FABRIC
+package io.github.null2264.cobblegen.mixin.create;
+
+import com.simibubi.create.api.event.PipeCollisionEvent;
+import com.simibubi.create.content.fluids.FluidReactions;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.state.BlockState;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import java.util.Optional;
+
+import static io.github.null2264.cobblegen.CobbleGen.FLUID_INTERACTION;
+
+/**
+ * Mixin for Create Fabric Patch F
+ */
+@Mixin(FluidReactions.class)
+public abstract class CreateFluidReactionsMixinPatchF
+{
+    @Inject(
+        method = "handlePipeFlowCollisionFallback",
+        at = @At(value = "HEAD"), cancellable = true
+    )
+    private static void generator$handlePipeFlowCollision(PipeCollisionEvent.Flow event, CallbackInfo ci) {
+        final Optional<BlockState> state =
+            FLUID_INTERACTION.interactFromPipeState(event.getLevel(), event.getPos(), event.getFirstFluid(), event.getSecondFluid());
+        if (state.isPresent()) {
+            event.setState(state.get());
+            ci.cancel();
+        }
+    }
+}
+//#endif

--- a/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchF.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchF.java
@@ -1,4 +1,4 @@
-//#if FABRIC
+//#if FABRIC && MC>1.16.5
 package io.github.null2264.cobblegen.mixin.create;
 
 import com.simibubi.create.api.event.PipeCollisionEvent;

--- a/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchF.java
+++ b/src/main/java/io/github/null2264/cobblegen/mixin/create/CreateFluidReactionsMixinPatchF.java
@@ -2,11 +2,9 @@
 package io.github.null2264.cobblegen.mixin.create;
 
 import com.simibubi.create.api.event.PipeCollisionEvent;
-import com.simibubi.create.content.fluids.FluidReactions;
-import net.minecraft.core.BlockPos;
-import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockState;
 import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Pseudo;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
@@ -18,7 +16,8 @@ import static io.github.null2264.cobblegen.CobbleGen.FLUID_INTERACTION;
 /**
  * Mixin for Create Fabric Patch F
  */
-@Mixin(FluidReactions.class)
+@Pseudo
+@Mixin(targets = {"com.simibubi.create.content.contraptions.fluids.FluidReactions", "com.simibubi.create.content.fluids.FluidReactions"})
 public abstract class CreateFluidReactionsMixinPatchF
 {
     @Inject(

--- a/src/main/resources/cobblegen.mixins.json
+++ b/src/main/resources/cobblegen.mixins.json
@@ -14,7 +14,7 @@
         "CommandsMixin",
         "MinecraftServerMixin",
         //#if MC>1.16.5
-        "fluid.CreateFluidReactionsMixin",
+        "create.CreateFluidReactionsMixin",
         //#endif
         "fluid.FluidEventMixin",
         "fluid.LavaEventMixin"

--- a/src/main/resources/cobblegen.mixins.json
+++ b/src/main/resources/cobblegen.mixins.json
@@ -16,7 +16,9 @@
         //#if MC>1.16.5
         "create.CreateFluidReactionsMixin",
         "create.CreateFluidReactionsMixinPatchE",
+        //#if FABRIC
         "create.CreateFluidReactionsMixinPatchF",
+        //#endif
         //#endif
         "fluid.FluidEventMixin",
         "fluid.LavaEventMixin"

--- a/src/main/resources/cobblegen.mixins.json
+++ b/src/main/resources/cobblegen.mixins.json
@@ -15,6 +15,8 @@
         "MinecraftServerMixin",
         //#if MC>1.16.5
         "create.CreateFluidReactionsMixin",
+        "create.CreateFluidReactionsMixinPatchE",
+        "create.CreateFluidReactionsMixinPatchF",
         //#endif
         "fluid.FluidEventMixin",
         "fluid.LavaEventMixin"

--- a/src/main/resources/fabric.mod.json
+++ b/src/main/resources/fabric.mod.json
@@ -50,8 +50,5 @@
     "depends": {
         "fabricloader": "*",
         "minecraft": "${mcversion}"
-    },
-    "breaks": {
-        "create": ">=0.5.1"
     }
 }

--- a/src/main/resources/fabric.mod.json
+++ b/src/main/resources/fabric.mod.json
@@ -50,5 +50,8 @@
     "depends": {
         "fabricloader": "*",
         "minecraft": "${mcversion}"
+    },
+    "breaks": {
+        "create": ">=0.5.1"
     }
 }

